### ISSUE TITLE
Add a new flag to `pulumi import` to allow running in pure codegen mode

### ISSUE THREAD

**Issue by Frassle** [pulumi/pulumi#8836]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Add a new flag (i.e. `--generate-only`, or `-g`; like `pulumi new`) that always outputs the code with `ResourceOptions.import` and doesn’t change the stack state itself.

Question: As an alternative, should we consider changing the default behavior of `pulumi import` to behave this way, and provide a new flag (i.e. `--auto`) to opt-in to the old default behavior?

**Comment by ProdigySim** [comment]
As a new user to pulumi who is migrating existing infrastructure, this flag would be helpful.

We would like to stick to using CICD to apply changes to our pulumi infrastructure, so we would prefer to use the `{ import: ID! }` property to import existing infra, but the auto-generated code has been extremely helpful in getting ourselves up to speed when prototyping and testing out Pulumi.

Being able to leave instructions for my devs to run `pulumi import --generate-only` and check in the code would really streamline the experience of using pulumi and keep it in line with normal CICD usage.

**Comment by dixler** [comment]
@mikhailshilkov is this completed?

**Comment by Frassle** [comment]
It is not

**Comment by adam-kiss-sg** [comment]
Adding a use-case:

We are migrating a multi-environment cdk stack to pulumi, where the environments (dev/prod) only differ in config. In this case it would be really nice to be able to run the pulumi import on the dev to generate the code, and replace the values with config until the import works for all envs. We use pulumi stacks for the different envs, so the "difference" between the envs can come from pulumi config.

**Comment by ProdigySim** [comment]
To solve this for my own developers, I put together a script that runs the import in a temporary stack, then tears it down.

Here's a simplified version of the script:
```bash
#!/bin/bash

RESOURCE_TYPE=$1
RESOURCE_ID=$2

pulumi stack init junk-stack

## import the resource and put generated code in a temporary file
pulumi import --yes --skip-preview -s junk-stack \
  --out TEMP_RESOURCE_CODE \
  $RESOURCE_TYPE \
  CHANGEME_UNUSED_NAME \
  $RESOURCE_ID

## Do a stack export to capture additional settings, save it in a file with the resource ID
pulumi stack export -s junk-stack > $RESOURCE_ID.json

## Remove the temporary stack without destroying resources
pulumi stack rm --force --yes -s junk-stack

cat TEMP_RESOURCE_CODE
rm TEMP_RESOURCE_CODE
```

Our real version has a lot more error handling and logging and such, but this is the broad strokes.

**Comment by adam-kiss-sg** [comment]
I think this is a duplicate of https://github.com/pulumi/pulumi/issues/8836

**Comment by a523** [comment]
yeah, I think this is a really import feature because sometimes writing code is harder than using the console to create a resource, We can generate basic code in this way.  To be honest, this was inspired by kubectl create deploy xxx --dry-run -o yaml I don’t think it would be too difficult to implement this feature on a current basis. Are you willing to accept open-source contributions? I want to try if possible

**Comment by warpedgeoid** [comment]
So, I take it this still hasn't been implemented two years later?

**Comment by bra-fsn** [comment]
> So, I take it this still hasn't been implemented two years later?

Doesn't seem to be.
The best I could get is to run the `import` with the `--diff` flag, but that prints a structured, text output instead of the actual code.
Still somewhat useful, but far from being as helpful as the code gen would be.

**Comment by adam-kiss-sg** [comment]
This is what I have in my .bashrc now, hope it helps:

```
function pulumiimport() {
  TYPE="$1"
  NAME="$2"
  ID="$3"

  STACK_NAME=$(pulumi stack --show-name)
  PROJECT_NAME=$(sed -rn 's/^name: ([a-z-]+)\s*$/\1/p' Pulumi.yaml)

  pulumi import "$TYPE" "$NAME" "$ID"

  pulumi state delete --force -y "urn:pulumi:$STACK_NAME::$PROJECT_NAME::$TYPE::$NAME"
}
```

**Comment by bra-fsn** [comment]
> This is what I have in my .bashrc now, hope it helps:
> 
> ```
> function pulumiimport() {
>   TYPE="$1"
>   NAME="$2"
>   ID="$3"
> 
>   STACK_NAME=$(pulumi stack --show-name)
>   PROJECT_NAME=$(sed -rn 's/^name: ([a-z-]+)\s*$/\1/p' Pulumi.yaml)
> 
>   pulumi import "$TYPE" "$NAME" "$ID"
> 
>   pulumi state delete --force -y "urn:pulumi:$STACK_NAME::$PROJECT_NAME::$TYPE::$NAME"
> }
> ```

Yeah, I would do something similar, but there are [other problems](https://github.com/pulumi/pulumi/issues/13339), which make this experience a little bit bumpy. :)

**Comment by sebpeterson** [comment]
My 2 cents, 
As a new user, this would be the best way for me to generate code from my existing infra, without committing to my stack.
I don't get why this is not possible, and is really slowing me down.

**Comment by rulatir** [comment]
Forget it. They'll never add it. It's one of those low hanging fruits they will keep finding excuses not to implement, keep users in the state of perpetual tantalization with ALMOST-usability, and keep IaC a forever "promising" but never delivering invention.

**Comment by rulatir** [comment]
> Gotta love finding a 3 and half year old ticket to do something you'd think would be super basic.

I know, right? It is not even a ticket to DO something. It is a ticket NOT to do something. To simply `if` around the entire bit where the state is commited.

**Comment by Frassle** [comment]
> To simply `if` around the entire bit where the state is commited.

I promise you if it were that simple I would have just done it. The current import mechanism is heavily tied into running a deployment and saving to state. It's relatively simple to change the code to just use a temp file based state store instead of the real state store but you still get all the UI showing you that it's doing a deployment, even though nothing is then saved. 
More importantly without further changes the code written out is missing the "importId" resource option, so given no actual state was changed if you just copy paste the code into your program it won't import it, it will create a new one.

Making this good requires a decent amount of effort.
A version that still runs a deployment and doesn't have the right import option you can do easily enough at home by just creating a temp stack, importing into it and then removing it, something like https://github.com/pulumi/pulumi/issues/8836#issuecomment-1198513778.
